src/main/java/org/stellar/sdk/Sep10Challenge.java

```diff
@@ -40,4 +44,336 @@ public static String newChallenge(
+   * @throws IOException                    If read XDR string fails, the exception will be thrown.
+   */
+  public static LinkedHashSet<String> verifyChallengeTransactionSigners(String challengeXdr, String serverAccountId, Network network, Set<String> signers) throws InvalidSep10ChallengeException, IOException {
+    if (signers == null || signers.isEmpty()) {
+      throw new InvalidSep10ChallengeException("No signers provided.");
+    }
+
+    // Read the transaction which validates its structure.
+    ChallengeTransaction parsedChallengeTransaction = readChallengeTransaction(challengeXdr, serverAccountId, network);
+    Transaction transaction = parsedChallengeTransaction.getTransaction();
+
+    // Ensure the server account ID is an address and not a seed.
+    KeyPair serverKeyPair = KeyPair.fromAccountId(serverAccountId);
+
+    // Deduplicate the client signers and ensure the server is not included
+    // anywhere we check or output the list of signers.
+    Set<String> clientSigners = new HashSet<String>();
+    for (String signer : signers) {
+      // Ignore non-G... account/address signers.
+      StrKey.VersionByte versionByte;
+      try {
+        versionByte = StrKey.decodedVersionByte(signer);
+      } catch (Exception e) {
+        continue;
+      }
+
+      if (!StrKey.VersionByte.ACCOUNT_ID.equals(versionByte)) {
+        continue;
+      }
+
+      // Ignore the server signer if it is in the signers list. It's
+      // important when verifying signers of a challenge transaction that we
+      // only verify and return client signers. If an account has the server
+      // as a signer the server should not play a part in the authentication
+      // of the client.
+      if (serverKeyPair.getAccountId().equals(signer)) {
+        continue;
+      }
+      clientSigners.add(signer);
+    }
+
+    // Don't continue if none of the signers provided are in the final list.
+    if (clientSigners.isEmpty()) {
+      throw new InvalidSep10ChallengeException("No verifiable signers provided, at least one G... address must be provided.");
+    }
+
+    // Verify all the transaction's signers (server and client) in one
+    // hit. We do this in one hit here even though the server signature was
+    // checked in the readChallengeTx to ensure that every signature and signer
+    // are consumed only once on the transaction.
+    Set<String> allSigners = new HashSet<String>(clientSigners);
+    allSigners.add(serverKeyPair.getAccountId());
+    LinkedHashSet<String> allSignersFound = verifyTransactionSignatures(transaction, allSigners);
+
+    // Confirm the server is in the list of signers found and remove it.
+    boolean serverSignerFound = allSignersFound.remove(serverKeyPair.getAccountId());
+    // After removing the server signer we call it clientSignersFound
+    LinkedHashSet<String> clientSignersFound = allSignersFound;
```

Comment:
Does this look strange?

Reply:
A little. I'd get rid of allSignersFound and just call it signersFound from the start.